daemon/cmd/daemon.go

```diff
@@ -276,7 +277,55 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 		fromFS = node.GetInternalIPv4Router()
 	}
 
-	node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	if err := removeOldRouterState(restoredIP); err != nil {
+		log.WithError(err).Warnf(
+			"Failed to remove old router IPs (restored IP: %s) from cilium_host. Manual intervention is required to remove all other old IPs.",
+			restoredIP,
+		)
+	}
+}
+
+// removeOldRouterState will try to ensure that the only IP assigned to the
+// `cilium_host` interface is the given restored IP. If the given IP is nil,
+// then it attempts to clear all IPs from the interface.
+func removeOldRouterState(restoredIP net.IP) error {
+	l, err := netlink.LinkByName(defaults.HostDevice)
+	if err != nil {
+		return err
+	}
+
+	family := netlink.FAMILY_V6
+	if restoredIP.To4() != nil {
+		family = netlink.FAMILY_V4
+	}
+	addrs, err := netlink.AddrList(l, family)
+	if err != nil {
+		return err
+	}
+	if len(addrs) > 1 {
```

Comment:
If `len(addrs) == 1` and `restoredIP` is nil, is the assumption correct that we keep the old router ip?

Reply:
If it is indeed the case, maybe a comment should be added to give some more context on why it is a valid case.

Reply:
Good point @borkmann. The [diff](https://github.com/cilium/cilium/compare/a9c54b38e38507339e8dca284b248080e7a92858..587f4c19a16b22c3e2a6018aba2f9b5b6ea9e3ad) of the change is here. It is not a correct to keep the old router IP because it is not valid. We therefore should remove it from the interface.